### ISSUE TITLE
Usage of OpenSSL 3.0 for Yocto kirkstone

### ISSUE THREAD

**Issue by SteffenRumler** [Azure/iot-identity-service#413]
I'm using the iot-identity-service 1.2.6.
I'd like to prepare switching from Yocto dunfell to Yocto kirkstone. This includes switching the OpenSSL version from `1.1.1` to` 3.0.x`.
I've tried to build the iot-identity-service using OpenSSL version `3.0.3`. But this fails.
Are there any plans to support OpenSSL version `3.0.x`?

---

Edit by @arsing - For quick reference, the current status is:

- OpenSSL 3.0 support has been implemented in [main.](https://github.com/Azure/iot-identity-service/commit/f66c155113cce5089d398376d16b206f7973d6f6)

- The OpenSSL library must have been compiled with support for the deprecated ENGINE API, as described in [this OpenSSL doc.](https://www.openssl.org/docs/man3.0/man7/migration_guide.html#Engines-and-METHOD-APIs) This is known to be true for Ubuntu 22.04's distro package.

- Support has not yet been released in a stable release of aziot-identity-service package.

**Comment by arsing** [comment]
Not any time soon.

3.0 changed the engine API completely, so adding support for it is a large effort. Also some of the existing distros we support will never have 3.0 support, so we'd have to support both in parallel instead of just switching entirely, which is even more effort.

That said, it's worth figuring out what exactly doesn't compile with 3.0 today. What I wrote above applies to the aziot-keys openssl engine specifically, but I haven't checked whether keyd itself, and the other i-i-s services, compile with 3.0 or not for their other uses of openssl (TLS, symmkey stuff, etc). The `openssl` crate itself does support 3.0 and with the same wrapped API, AFAIK.

Since you've already tried, could you post the log?

**Comment by arsing** [comment]
Well, the fact that libaziot-keys itself doesn't build prevents us from knowing what else is broken in the services, but at the very least here's the list of other things that need to be changed to get to that point:

- `openssl-errors` dep needs to be updated from 0.1 to 0.2

- `cargo update -p openssl -p openssl-sys -p openssl-build`

- The `ENGINE_get_ex_new_index` / `EC_KEY_get_ex_new_index` / `RSA_get_ex_new_index` functions have been fixed to use the correct signature for the dup function (`from_d` parameter is now `void**` instead of `void*`), so C declaration and Rust declaration need to be updated to match.

```diff,c
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+int aziot_key_dupf_engine_ex_data(CRYPTO_EX_DATA *to, const CRYPTO_EX_DATA *from, void **from_d, int idx, long argl, void *argp);
+#elif OPENSSL_VERSION_NUMBER >= 0x10100000L
 int aziot_key_dupf_engine_ex_data(CRYPTO_EX_DATA *to, const CRYPTO_EX_DATA *from, void *from_d, int idx, long argl, void *argp);
 #else
 int aziot_key_dupf_engine_ex_data(CRYPTO_EX_DATA *to, CRYPTO_EX_DATA *from, void *from_d, int idx, long argl, void *argp);
 #endif
```

```diff,rust
 #[no_mangle]
 #[allow(clippy::similar_names)]
 unsafe extern "C" fn aziot_key_dupf_engine_ex_data(
     _to: *mut openssl_sys::CRYPTO_EX_DATA,
     _from: *const openssl_sys::CRYPTO_EX_DATA,
-    from_d: *mut std::ffi::c_void,
+    from_d: *mut *mut std::ffi::c_void,
     idx: std::os::raw::c_int,
     _argl: std::os::raw::c_long,
     _argp: *mut std::ffi::c_void,
 ) -> std::os::raw::c_int {
     crate::ex_data::dup::<openssl_sys::ENGINE, Engine>(from_d, idx);
     1
 }
```

```diff,rust
-pub(crate) unsafe fn dup<T, U>(from_d: *mut std::ffi::c_void, idx: std::os::raw::c_int)
+pub(crate) unsafe fn dup<T, U>(from_d: *mut *mut std::ffi::c_void, idx: std::os::raw::c_int)
 where
     T: HasExData<U>,
 {
     let ex_index = <T as HasExData<U>>::index().as_raw();
     assert_eq!(idx, ex_index);
 
-    // Although `dup_func`'s signature types `from_d` as `void*`, it is in fact a `void**` - it points to the pointer returned by
-    // calling `CRYPTO_get_ex_data` on the `from` object. After `dup_func` returns, openssl takes whatever `from_d` is pointing to,
+    // `from_d` points to the pointer returned by calling `CRYPTO_get_ex_data` on the `from` object.
+    // After `dup_func` returns, openssl takes whatever `from_d` is pointing to,
     // and sets it as the ex data of the `to` object using `CRYPTO_set_ex_data`.
```

- utpm and c-shared don't compile. onalante-msft is working on removing utpm anyway, so that will help with this as a side-effect.

**Comment by SteffenRumler** [comment]
I've prepared this patch, which makes the build possible: [0001-adapt-for-yocto-kirkstone.zip](https://github.com/Azure/iot-identity-service/files/8769790/0001-adapt-for-yocto-kirkstone.zip)

Basically it does the following:
- update crates openssl-errors, openssl, openssl-sys (and others)
- workaround build problems of C parts caused by check against magic number: OPENSSL_VERSION_NUMBER < 0x20000000L
- workaround build problems of C parts caused by gcc toolchain update 9.3.0 (dunfell) -> 11.2.0 (kirkstone): -Wno-array-parameter -Wno-deprecat     ed-declarations

I don't think this is complete (I'm not a rust expert). Trying it ends up in a TLS handshake error.

**Comment by arsing** [comment]
Thanks.

The `-Wno-incompatible-pointer-types` part is better done by using the right signature in the first place, ie the diff in my previous comment.

So the relevant part of the diff is (inlined for convenience):

<details>
<summary>aziot-tpm-sys.patch</summary>

```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7bbfa6f3..9d1989b0 100644
--- a/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/c-shared/CMakeLists.txt
+++ b/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/c-shared/CMakeLists.txt
@@ -176,7 +176,7 @@ elseif(LINUX) #LINUX OR APPLE
     endif()
     # Turn off warning that can not be fixed right now
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-variable -Wno-missing-braces -Wno-missing-field-initializers -Wno-format-nonliteral")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-unused-variable -Wno-missing-braces -Wno-missing-field-initializers -Wno-format-nonliteral")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-unused-variable -Wno-missing-braces -Wno-missing-field-initializers -Wno-format-nonliteral -Wno-array-parameter -Wno-deprecated-declarations")
 elseif(APPLE)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wformat-security")
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wformat-security")

diff --git a/adapters/tlsio_openssl.c b/adapters/tlsio_openssl.c
index a73c129c..c621c861 100644
--- a/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/c-shared/adapters/tlsio_openssl.c
+++ b/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/c-shared/adapters/tlsio_openssl.c
@@ -866,7 +866,7 @@ static int add_certificate_to_store(TLS_IO_INSTANCE* tls_io_instance, const char
         }
         else
         {
-#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER < 0x20000000L)
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER <= 0x30000030L)
             const BIO_METHOD* bio_method;
 #else
             BIO_METHOD* bio_method;

diff --git a/adapters/x509_openssl.c b/adapters/x509_openssl.c
index 11be7a03..96993f40 100644
--- a/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/c-shared/adapters/x509_openssl.c
+++ b/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/c-shared/adapters/x509_openssl.c
@@ -73,7 +73,7 @@ static int load_certificate_chain(SSL_CTX* ssl_ctx, const char* certificate)
                 // certificates.
 
                 /* Codes_SRS_X509_OPENSSL_07_006: [ If successful x509_openssl_add_ecc_credentials shall to import each certificate in the cert chain. ] */
-#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER < 0x20000000L)
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER <= 0x30000030L)
                 SSL_CTX_clear_extra_chain_certs(ssl_ctx);
 #else
                 if (ssl_ctx->extra_certs != NULL)
@@ -264,7 +264,7 @@ int x509_openssl_add_certificates(SSL_CTX* ssl_ctx, const char* certificates)
         else
         {
             /*Codes_SRS_X509_OPENSSL_02_012: [ x509_openssl_add_certificates shall get the memory BIO method function by calling BIO_s_mem. ]*/
-#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER < 0x20000000L)
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER <= 0x30000030L)
             const BIO_METHOD* bio_method;
 #else
             BIO_METHOD* bio_method;
-- 
2.30.2

diff --git a/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/CMakeLists.txt b/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/CMakeLists.txt
index 7bbfa6f3..9d1989b0 100644
--- a/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/CMakeLists.txt
+++ b/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/CMakeLists.txt
@@ -176,7 +176,7 @@ elseif(LINUX) #LINUX OR APPLE
     endif()
     # Turn off warning that can not be fixed right now
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-variable -Wno-missing-braces -Wno-missing-field-initializers -Wno-format-nonliteral")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-unused-variable -Wno-missing-braces -Wno-missing-field-initializers -Wno-format-nonliteral")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-unused-variable -Wno-missing-braces -Wno-missing-field-initializers -Wno-format-nonliteral -Wno-array-parameter -Wno-deprecated-declarations")
 elseif(APPLE)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wformat-security")
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wformat-security")

diff --git a/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/adapters/tlsio_openssl.c b/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/adapters/tlsio_openssl.c
index a73c129c..c621c861 100644
--- a/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/adapters/tlsio_openssl.c
+++ b/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/adapters/tlsio_openssl.c
@@ -866,7 +866,7 @@ static int add_certificate_to_store(TLS_IO_INSTANCE* tls_io_instance, const char
         }
         else
         {
-#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER < 0x20000000L)
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER <= 0x30000030L)
             const BIO_METHOD* bio_method;
 #else
             BIO_METHOD* bio_method;

diff --git a/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/adapters/x509_openssl.c b/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/adapters/x509_openssl.c
index 11be7a03..96993f40 100644
--- a/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/adapters/x509_openssl.c
+++ b/tpm/aziot-tpm-sys/azure-iot-hsm-c/deps/utpm/deps/c-utility/adapters/x509_openssl.c
@@ -73,7 +73,7 @@ static int load_certificate_chain(SSL_CTX* ssl_ctx, const char* certificate)
                 // certificates.
 
                 /* Codes_SRS_X509_OPENSSL_07_006: [ If successful x509_openssl_add_ecc_credentials shall to import each certificate in the cert chain. ] */
-#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER < 0x20000000L)
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER <= 0x30000030L)
                 SSL_CTX_clear_extra_chain_certs(ssl_ctx);
 #else
                 if (ssl_ctx->extra_certs != NULL)
@@ -264,7 +264,7 @@ int x509_openssl_add_certificates(SSL_CTX* ssl_ctx, const char* certificates)
         else
         {
             /*Codes_SRS_X509_OPENSSL_02_012: [ x509_openssl_add_certificates shall get the memory BIO method function by calling BIO_s_mem. ]*/
-#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER < 0x20000000L)
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER <= 0x30000030L)
             const BIO_METHOD* bio_method;
 #else
             BIO_METHOD* bio_method;
-- 
2.30.2
```

**Comment by arsing** [comment]
Is it using symmkey provisioning or X.509?

**Comment by SteffenRumler** [comment]
I'm using manual provisioning with a connection string:
```
# cat /etc/aziot/config.toml
[provisioning]
source = "manual"
# copy/paste 'Primary Connection String'
connection_string = "HostName=iothub-ics-dev.azure-devices.net;DeviceId=...;SharedAccessKey=..."
```

BTW, the TLS handshake using `openssl s_client` looks working:
```
# openssl s_client -connect iothub-ics-dev.azure-devices.net:443
CONNECTED(00000003)
depth=2 C = IE, O = Baltimore, OU = CyberTrust, CN = Baltimore CyberTrust Root
verify return:1
depth=1 C = US, O = Microsoft Corporation, CN = Microsoft RSA TLS CA 01
verify return:1
depth=0 CN = *.azure-devices.net
verify return:1
---
Certificate chain
 0 s:CN = *.azure-devices.net
   i:C = US, O = Microsoft Corporation, CN = Microsoft RSA TLS CA 01
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: May 25 07:46:20 2022 GMT; NotAfter: Oct 25 07:46:20 2022 GMT
 1 s:C = US, O = Microsoft Corporation, CN = Microsoft RSA TLS CA 01
   i:C = IE, O = Baltimore, OU = CyberTrust, CN = Baltimore CyberTrust Root
   a:PKEY: rsaEncryption, 4096 (bit); sigalg: RSA-SHA256
   v:NotBefore: Jul 21 23:00:00 2020 GMT; NotAfter: Oct  8 07:00:00 2024 GMT
---
Server certificate
-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----
subject=CN = *.azure-devices.net
issuer=C = US, O = Microsoft Corporation, CN = Microsoft RSA TLS CA 01
---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA
Server Temp Key: ECDH, prime256v1, 256 bits
---
SSL handshake has read 3915 bytes and written 504 bytes
Verification: OK
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES128-SHA256
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES128-SHA256
    Session-ID: ...
    Session-ID-ctx: 
    Master-Key: ...
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1653911940
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: yes
---
^C
```

**Comment by arsing** [comment]
Okay, if symmkey provisioning fails then that *ought* to rule out any problem with the engine.

We create openssl client connectors in http-common/src/proxy.rs `fn make_tls_connector`. For symmkey provisioning `identity` will be `None`, so the only relevant stuff is that the connector is created using `openssl::ssl::SslConnector::builder(openssl::ssl::SslMethod::tls())`. From the openssl crate docs that seems to still be fine with openssl 3.0.

It could be that openssl is trying to do something with the engine anyway (with the server cert's key?) *and* that isn't working for some reason. You'd need to debug it to know.

**Comment by micahl** [comment]
> Are there any plans to support OpenSSL version `3.0.x`?

Yes, but not in the very near term.  It is something we know we'll have to address given that Ubuntu 22.04 has also switched to openssl 3.0.

**Comment by JanZachmann** [comment]
> > Are there any plans to support OpenSSL version `3.0.x`?
> 
> Yes, but not in the very near term. It is something we know we'll have to address given that Ubuntu 22.04 has also switched to openssl 3.0.

not only Ubuntu 22.04 introduces a dependency to OpenSSl 3.0.x.  The new Yocto LTS release Kirkstone also requires openssl 3.0

**Comment by mlilien** [comment]
> May 30 07:46:20 test aziot-identityd[1832]: 2022-05-30T07:46:20Z [WARN] - Failed to communicate with IoT Hub (attempt 1 of 1): error trying to connect: error:0A080006:SSL routines:ssl_generate_param_group:EVP lib:../openssl-3.0.3/ssl/s3_lib.c:4747:, error:0A00013A:SSL routines:tls_process_ske_ecdhe:unable to find ecdh parameters:../openssl-3.0.3/ssl/statem/statem_clnt.c:2144: 

Applications  linked against openssl 3.0 have to at least use the default provider (https://github.com/openssl/openssl/blob/master/README-PROVIDERS.md). 
After doing that, the above error is gone.  I tested kirkstone with iot-identity-service 1.3.0, rust 1.61.0 and symmetric key and x509 provisioning as well. I did not test tpm provisioning yet.

**Comment by Dvergatal** [comment]
> > May 30 07:46:20 test aziot-identityd[1832]: 2022-05-30T07:46:20Z [WARN] - Failed to communicate with IoT Hub (attempt 1 of 1): error trying to connect: error:0A080006:SSL routines:ssl_generate_param_group:EVP lib:../openssl-3.0.3/ssl/s3_lib.c:4747:, error:0A00013A:SSL routines:tls_process_ske_ecdhe:unable to find ecdh parameters:../openssl-3.0.3/ssl/statem/statem_clnt.c:2144:
> 
> Applications linked against openssl 3.0 have to at least use the default provider (https://github.com/openssl/openssl/blob/master/README-PROVIDERS.md). After doing that in the application explicitly, the above error is gone. I tested kirkstone with iot-identity-service 1.3.0, rust 1.61.0 and symmetric key and x509 provisioning as well. I did not test tpm provisioning yet.

In our case we have verified only tpm provisioning and it is working without the need for setting this provider. Also you have to get in account that ecdh is based on ecc and by default tpm provisioning uses rsa. I don't even know if azure supports ecc, probably not.

**Comment by arsing** [comment]
Since we have an unofficial working solution, I've edited the OP to link to the comment by Micah that we haven't planned to work on it at this time, and to the comment with the patches, so that users landing on this thread can find them easily.

**Comment by arsing** [comment]
>onalante-msft is working on removing utpm anyway, so that will help with this as a side-effect.

(This was done by https://github.com/Azure/iot-identity-service/commit/143cdf79be08e7069a6a0b17b38f33469c291b1e, ie the aziot-tpm-sys.patch is no longer needed for latest main.)

**Comment by Dvergatal** [comment]
> > onalante-msft is working on removing utpm anyway, so that will help with this as a side-effect.
> 
> (This was done by [143cdf7](https://github.com/Azure/iot-identity-service/commit/143cdf79be08e7069a6a0b17b38f33469c291b1e), ie the aziot-tpm-sys.patch is no longer needed for latest main.)

I have seen that in 1.4.0 there is this behaviour with the hierarchy tpm fixed:
>Use of the TCG TPM2 Software Stack which enables TPM hierarchy authorization values, specifying the TPM index at which to persist the DPS authentication key, and accommodating more TPM configurations ([doc](https://github.com/Azure/iotedge/blob/897aed8c5573e8cad4b602e5a1298bdc64cd28b4/edgelet/contrib/config/linux/template.toml#L262-L288))

I have a question because in #2329 it has been discussed that it should be fixed and now we comment above it sounds like it has been fixed and @RonaldAi doesn't have a clue because he hasn't been involved, so my confusion is, has it been already fixed and iotedge is now working properly with all the verifications that @williamcroberts mentioned about?

**Comment by arsing** [comment]
@Dvergatal Please open a new issue. This one is about OpenSSL 3.0 support.

**Comment by Dvergatal** [comment]
@arsing because you were writing about supporting this tss API instead of utpm I've assumed that I can continue this thread in here, but You asked me so kind, that I have opened a new issue :]

**Comment by ElD4n1** [comment]
Is anyone here willing to share the recipe for iot-identity-service with successful build?
I am currently trying around with kirkstone, iot-identity-service 1.3.0, (meta-)rust 1.61.0 and the above patches with some modifications, but keep running into build issues.

**Comment by ElD4n1** [comment]
> Added changes to the `main` branch for Ubuntu 22.04 that seem like they'd satisfy the ask here as well. Anyone targeting Yocto that could confirm? The current effort on a Yocto recipe in [Azure/meta-iotedge#90](https://github.com/Azure/meta-iotedge/pull/90) is targeting Dunfell (maybe not that different from Kirkstone)?

Yes, with the current main branch (commit 5994d1a6cc7640c2b53928ba9d5d574a3380dbfc) the aziotd package compiles for me without OpenSSL specific patches on Yocto kirkstone. The only patches that are still needed seem to be the same as in the mentioned draft for dunfell over here: https://github.com/Azure/meta-iotedge/pull/90. I used my own recipe though.

Tested using TPM provisioning. For this I also needed to add `libtss2 libtss2-mu libtss2-tcti-device` to RDEPENDS.

**Comment by micahl** [comment]
@EID4n1 thanks for trying it out!  Tagging @ggjjj for awareness on the RDEPENDS additions needed for TPM (if not already).

**Comment by micahl** [comment]
Closing this issue now that the [meta-iotedge's main](https://github.com/azure/meta-iotedge) has been updated for Kirkstone.
